Framework/Crystal/src/FindSXPeaksHelper.cpp

```diff
@@ -475,33 +504,85 @@ AllPeaksStrategy::getAllPeaks(const HistogramData::HistogramX &x, const Histogra
+NSigmaPeaksStrategy::NSigmaPeaksStrategy(const API::SpectrumInfo &spectrumInfo, const double nsigma,
+                                         const double minValue, const double maxValue, const XAxisUnit units)
+    : PeakFindingStrategy(nullptr, spectrumInfo, minValue, maxValue, units), m_nsigma(nsigma) {}
+
+PeakList NSigmaPeaksStrategy::dofindSXPeaks(const HistogramData::HistogramX &x, const HistogramData::HistogramY &y,
+                                            const HistogramData::HistogramE &e, Bound low, Bound high,
                                             const int workspaceIndex) const {
-  PeakList peaks;
+  auto nsigmaPeaks = getAllNSigmaPeaks(x, y, e, low, high);
+  auto sxPeaks = convertToSXPeaks(x, y, nsigmaPeaks, workspaceIndex);
+  return sxPeaks;
+}
 
-  if (foundPeaks.empty()) {
-    return peaks;
+std::vector<std::unique_ptr<PeakContainer>> NSigmaPeaksStrategy::getAllNSigmaPeaks(const HistogramData::HistogramX &x,
+                                                                                   const HistogramData::HistogramY &y,
+                                                                                   const HistogramData::HistogramE &e,
+                                                                                   Bound low, Bound high) const {
+
+  bool isRecording = false;
+  std::unique_ptr<PeakContainer> currentPeak = nullptr;
+  std::vector<std::unique_ptr<PeakContainer>> peaks;
+
+  // We want to the upper boundary to be inclusive hence we need to increment by one
+  if (high != x.end()) {
+    ++high;
   }
+  auto distanceMin = std::distance(x.begin(), low);
+  auto distanceMax = std::distance(x.begin(), high);
 
-  // Add a vector to the boost optional
-  peaks = std::vector<FindSXPeaksHelper::SXPeak>();
+  const auto lowY = y.begin() + distanceMin;
+  auto highY = distanceMax < static_cast<int>(y.size()) ? y.begin() + distanceMax : y.end();
 
-  for (const auto &peak : foundPeaks) {
-    // Get the index of the bin
-    auto maxY = peak->getMaxIterator();
-    const auto distance = std::distance(y.begin(), maxY);
-    const auto xValue = getXValue(x, distance);
-    const auto tof = convertToTOF(xValue, workspaceIndex);
-    const double phi = calculatePhi(workspaceIndex);
+  const auto lowE = e.begin() + distanceMin;
+  const auto highE = distanceMax < static_cast<int>(e.size()) ? e.begin() + distanceMax : e.begin();
 
-    std::vector<int> specs(1, workspaceIndex);
-    (*peaks).emplace_back(tof, phi, *maxY, specs, workspaceIndex, m_spectrumInfo);
+  // std::vector<double> yDifferences(distanceMax-distanceMin);
+  // std::transform(lowY+1, highY, lowY, yDifferences.begin(), std::minus<double>());
+
+  auto yIt = lowY + 1;
+  auto eIt = lowE + 1;
+
+  for (; yIt != highY && eIt != highE; ++yIt, ++eIt) {
+    const auto signalDiff = *yIt - *(yIt - 1);
+    const auto isStartPeak = signalDiff > m_nsigma * (*eIt);
```

Comment:
Do we need to consider the special case of comparing two empty bins - could this floating point comparison ever be True, just due to precision?

Reply:
Do you mean the case when either `*yIt` or `*eIt` is empty? 
Can `>` operator has an impact due to double precision? Because during testing I have seen both True and False cases coming.

Reply:
If `*eIt == 0 ` then `*yIt ==0` no?
Yes the question is can `0 > 0` return `True` due to floating point precision - should there be some sort of tolerance?